compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1686,30 +1707,98 @@ private BType getListConstructorCompatibleNonUnionType(BType type) {
-        } else if ((arrayType.state != BArrayState.OPEN) && (arrayType.size != listConstructor.exprs.size())) {
-            if (arrayType.size < listConstructor.exprs.size()) {
-                dlog.error(listConstructor.pos,
-                        DiagnosticErrorCode.MISMATCHING_ARRAY_LITERAL_VALUES, arrayType.size,
-                        listConstructor.exprs.size());
+        } else if (arrayType.state != BArrayState.OPEN && arrayType.size != listExprSize) {
+            if (arrayType.size < listExprSize) {
+                dlog.error(listConstructor.pos, DiagnosticErrorCode.MISMATCHING_ARRAY_LITERAL_VALUES, arrayType.size,
+                        listExprSize);
                 return symTable.semanticError;
             }
 
             if (!types.hasFillerValue(eType)) {
-                dlog.error(listConstructor.pos,
-                        DiagnosticErrorCode.INVALID_LIST_CONSTRUCTOR_ELEMENT_TYPE, expType);
+                dlog.error(listConstructor.pos, DiagnosticErrorCode.INVALID_LIST_CONSTRUCTOR_ELEMENT_TYPE, expType);
                 return symTable.semanticError;
             }
         }
 
         boolean errored = false;
         for (BLangExpression expr : listConstructor.exprs) {
-            if (exprIncompatible(eType, expr) && !errored) {
-                errored = true;
+            if (expr.getKind() != NodeKind.LIST_CONSTRUCTOR_SPREAD_OP) {
+                errored |= exprIncompatible(eType, expr);
+                continue;
+            }
+
+            BLangExpression spreadOpExpr = ((BLangListConstructorSpreadOpExpr) expr).expr;
+            BType spreadOpType = checkExpr(spreadOpExpr, this.env);
+            spreadOpType = Types.getReferredType(spreadOpType);
+
+            switch (spreadOpType.tag) {
+                case TypeTags.ARRAY:
+                    BType spreadOpeType = ((BArrayType) spreadOpType).eType;
+                    if (types.typeIncompatible(spreadOpExpr.pos, spreadOpeType, eType)) {
+                        return symTable.semanticError;
+                    }
+                    break;
+                case TypeTags.TUPLE:
+                    BTupleType spreadOpTuple = (BTupleType) spreadOpType;
+                    List<BType> tupleTypes = spreadOpTuple.tupleTypes;
+                    for (BType tupleMemberType : tupleTypes) {
+                        if (types.typeIncompatible(spreadOpExpr.pos, tupleMemberType, eType)) {
+                            return symTable.semanticError;
+                        }
+                    }
+
+                    if (!isFixedLengthTuple(spreadOpTuple)) {
+                        if (types.typeIncompatible(spreadOpExpr.pos, spreadOpTuple.restType, eType)) {
+                            return symTable.semanticError;
+                        }
+                    }
+                    break;
+                default:
+                    dlog.error(spreadOpExpr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES_LIST_SPREAD_OP, spreadOpType);
```

Comment:
Use the refernce type for the error

Reply:
Not sure if I get this. We already do `spreadOpType = Types.getReferredType(spreadOpType);` before coming here?

Reply:
```ballerina
type Foo int;

function dummyTest() {
    Foo n = 10;
    int[] _ = [1, 2, ...n]; //atm error; incompatible types: expected an array or a tuple, found 'int'
}
```

Instead, the error should say **_incompatible types: expected an array or a tuple, found 'Foo'_**